posts/crowdsales-and-multiple-inheritance.md

```diff
@@ -0,0 +1,327 @@
+means that `super` on `D` calls `C`. And you might be a little surprised by the
+fact that calling `super` on `C` will result on a call to `B` instead of `A`,
+even though `C` doesn't inherit from `B`. Finally, `super` on `B` will call
+`A`.
+
+If `D` is instead defined as:
+
+```
+contract D is C, B {}
+```
+
+then `D.f` will call:
+
+```
+somethingD();
+somethingB();
+somethingC();
+somethingA();
+```
+
+When `D` inherits from `C, B`, the linearization results in D→ B→ C→ A.
+
+Notice here that the order in which you declare the parent contracts matters a
+lot. If the inheritance graph is not too complex, it will be easy to see that
+the order of calls will follow the order in which the parents were declared,
+right to left. If your inheritance is more complicated than this and the
+hierarchy is not very clear, you should probably stop using multiple
+inheritance and look for an alternate solution.
+
+I recommend you to read the wikipedia pages of
+[Multiple inheritance](https://en.wikipedia.org/wiki/Multiple_inheritance) and
+[C3 linearization](https://en.wikipedia.org/wiki/C3_linearization), and the
+[Solidity docs about multiple inheritance](https://solidity.readthedocs.io/en/latest/contracts.html#multiple-inheritance-and-linearization).
+You will find in there a complete explanation of the C3 algorithm, and an
+example with a more complicated inheritance graph.
+
+To better understand how this can impact your crowdsales, take a look at the
+case that Philip Daian brilliantly explains on his blog post
+[Solidity anti-patterns: Fun with inheritance DAG abuse](https://pdaian.com/blog/solidity-anti-patterns-fun-with-inheritance-dag-abuse/).
+There, he presents a Crowdsale contract that needs "to have a whitelist
+pool of preferred investors able to buy in the pre-sale [...], along with a
+hard cap of the number of [...] tokens that can be distributed." On his first
+(deliberately) faulty implementation, he ends up with a crowdsale that checks:
+
+```
+((withinPeriod && nonZeroPurchase) && withinCap) || (whitelist[msg.sender] && !hasEnded())
+```
+
+Pay close attention to the resulting condition, and notice that if a
+whitelisted investor buys tokens before the crowdsale has ended, she will be
+able to bypass the hard cap, and buy as many tokens as she wants.
+
+By just inverting the order on which the parents are defined, he fixes the
+contract which will now check for:
+
+```
+(withinPeriod && nonZeroPurchase || (whitelist[msg.sender] && !hasEnded())) && withinCap
+```
+
+Now, if the purchase attempt goes above the cap, the transaction will be
+reverted even if the sender is whitelisted.
+
+There are some simple cases, where the order of the conditions is not
+important. By now, you should have started to suspect that the contract above
+became complicated because of the `||` (or) condition. Things are a lot easier
+when all our conditions are merged with `&&` (and), because in that case the
+order of the conditions doesn't alter the result.
```

Comment:
Perhaps mention that our crowdsale architecture was crafted for using requires instead of returning booleans. So instead of providing evaluations that can be incorrectly concatenated with ||'s, all that evaluations can do are either revert() or do nothing.